### PR TITLE
add support for internal groups

### PR DESCRIPTION
**Pull request by dwelch0**
Adds functionality to create, update, and delete vault groups. Groups are based off role definitions referenced by data/teams/user definitions. The applicable roles must contain `oidc_permissions:` block with references to permissions that include `service: vault` in order for the group to be created. Users (entities) will be added to the vault groups that map to the applicable roles referenced in their user definition files.

tickets: https://issues.redhat.com/browse/APPSRE-4658

### CODE REVIEW

**Comment by dwelch0** [comment]
Note: branched off entity-refactor. merge entity pr before this

File: toplevel/group/group.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+		}
+		for _, u := range toBeUpdated {
+			u.(group).CreateOrUpdate("updated")
+		}
+	}
+}
+
+// processDesired accepts the yaml-marshalled result of the `vault_groups` graphql
+// query and returns group objects
+func processDesired(entries []user, entityNamesToIds map[string]string) []group {
```
[the remark this answers is not in the document]
Comment (dwelch0) [reply toplevel/group/group.go:135]: Agreed. I felt slightly handcuffed into writing it this way due to the multi layer relation from user to policy that must be traversed. Abstracting some nested logic intoto separate calls would certainly help. I will get on it

**Comment by bkez322** [comment]
Otherwise... Looks great to me.

File: toplevel/group/group.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+		"[Vault Identity] group successfully deleted")
+}
+
+var _ vault.Item = group{}
+
+func init() {
+	toplevel.RegisterConfiguration("vault_groups", config{})
+}
+
+func (c config) Apply(entriesBytes []byte, dryRun bool, threadPoolSize int) {
```
Comment (bkez322) [inline toplevel/group/group.go:100]: Considering how much it has to do, I love how clean and easy-to-follow this function is as written. 👍

Comment (dwelch0) [reply toplevel/group/group.go:100]: glad to hear it. It was in pretty rough shape for awhile :sweat_smile:

**Comment by TGPSKI** [comment]
rebase

**Review by TGPSKI — APPROVED** [review]

**Pull request merged**